### DISCUSSION TITLE
`RestRserve` `router` end point as a module

### DISCUSSION THREAD

**Discussion by dereckmezquita** [rexyai/RestRserve#203] [Q&A] [answered]
TLDR: how to create a router add an end point to it with the logic, and then add it to the `RestRserve` application. The purpose is to modularise my code. Finally, this would allow me to group my routes and then import that router into app.

I'm looking to do something like this:

- index.R: app
- user_data.R: user_data_router (groups routes)
    - user_data_route_1.R
    - user_data_route_2.R

So in index.R I would mount/add a single router, which in turn groups together other routes (by routes I mean end points)

I'm learning `RestRserve` I didn't find this online nor in the docs. I'm looking to create a `router` object and add an endpoint to it in a separate file, then export it so I can import from my main.R and add it to my application. The purpose is to modularise my code.

My minimal example:

```r
#!/usr/bin/env Rscript
box::use(RestRserve)

app <- RestRserve$Application$new()

app$add_get("/echo", \(.req, .res) {
    .res$set_body(list(paste0("The message is: '", .req$get_param_query("msg"), "'")))
    .res$set_content_type("application/json")
})


backend <- RestRserve$BackendRserve$new()
backend$start(app, http_port = 8080)
```

I was hoping in a separate file called `echo.R` I can do the following (this does not work; pseudocode):

```r
box::use(RestRserve)

echo <- RestRserve$Router$new()

echo$add_get("/echo", \(.req, .res) {
    .res$set_body(list(paste0("The message is: '", .req$get_param_query("msg"), "'")))
    .res$set_content_type("application/json")
})

#' @export
echo <- echo
```

`RestRserve` does indeed have a `Router` class but the documentation is hidden and does not have any methods for adding end points.

If I can do this then I would be able to import into my main.R and use it as such I hope:

```r
box::use(RestRserve)

app <- RestRserve$Application$new()

box::use(./echo[ echo ])

app$mount("/", echo)

backend <- RestRserve$BackendRserve$new()
backend$start(app, http_port = 8080)
```

I know this can be done using `plumber` I don't want to use `plumber`.

**Comment by dereckmezquita** [comment, marked as answer]
I was able to make something work, but I feel this is a hack. I would appreciate input from the developers of `RestRserve` for an official way of creating these modules, @dselivanov.

I would like to see the ability to create a `Router` and import that into `Application`. I will continue reading the source code to see if I can find any better way.

main.R
```r
#!/usr/bin/env Rscript
box::use(RestRserve)

app <- RestRserve$Application$new()

box::use(./user_routes/index[ user_routes ])

user_routes(app)

backend <- RestRserve$BackendRserve$new()
backend$start(app, http_port = 8080)
```

user_routes/index.R
```r
box::use(RestRserve)

box::use(./user_route_1[ user_route_1 ])
box::use(./user_route_2[ user_route_2 ])

#' @export
user_routes <- function(app) {
    user_route_1(app)
    user_route_2(app)

    invisible(app)
}
```
user_routes/user_route_1.R
```r
box::use(RestRserve)

#' @export
user_route_1 <- function(app) {
    app$add_get("/user_route_1", function(.req, .res) {
        .res$set_body("This is user route 1.")
        .res$set_content_type("text/plain")
    })
}
```
user_routes/user_route_2.R
```r
box::use(RestRserve)

#' @export
user_route_2 <- function(app) {
    app$add_get("/user_route_2", function(.req, .res) {
        .res$set_body("This is user route 2.")
        .res$set_content_type("text/plain")
    })
}
```
